### ISSUE TITLE
DSA pqgVer: spurious error

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP#439]
After completing a full run of all cipher implementations for OpenSSL, I get passing verdicts everywhere, except one single instance of one individual test in a DSA pqgVer.

vsID 18588, tcId 3.

I have uploaded pqgVer OpenSSL results for other vsIDs which all pass. When I rerun the IUT testing irrespective of the used cipher implementation, the IUT always generates the same result for tcId 3.

As for all other pqgVer tests I get passing results, I am now very unsure what is going on. The log of the ACVP server is not helpful:

```
    }, {
      "tcId" : 3,
      "result" : "fail",
      "reason" : "none"
```

May I ask you for a hint why the ACVP server does not like the tcId? Unfortunately I do not have another IUT that has a probable pqgVer implementation so that I can cross-check.

Thanks a lot.

**Comment by celic** [comment]
Ah. The `"none"` here means that the server did not introduce an error to that particular p,q,g combination. So the server expected the client to return `"testPassed": true` indicating that the server-provided p,q,g was successfully verified by the client. This result would mean the client returned `"testPassed": false` which the server thinks is incorrect. Let me look at the files and see if anything jumps out at me.

**Comment by smuellerDD** [comment]
After debugging, I get total different P, Q and counter values with the given domain seed than provided with the test vector. I get the following values:

Original P:
P = 9a0e05304f56e68d403e5ee54a2e2b7f1bdb36850f18a62
a6151dcf19cef4806b1915cf212330a3491415b078b98fbe9805bd0453bb95a0abc3fbaeb60ad58f
ec7ad1c7b9e63d9b33dd85e44b172e3b476fa8e50f5c671368f20a792e614771cb7967c300eb625b
415294c3f4ee71724beaa3d18b9d9d27ab226ece4adc61c3e1ffff508df747d03aefbc862a95bcb5
5f7bed7b9facaf0a4c151d76d95a5ecb16f41c4a7c4147f37687ff6202afa035ca5654329b777b1c
4c8f1499abf7f01713b7d3714e89c4c4fe7df7658d95a95a654047d4d5f39197c3ad8fdb01a4bf89
67942e592e9bc9fa98a7d2b2d048413362d7e5b9573b7bc1e359bc36ad75b248f

Generated P:
gen P = 90d9f0b7ed3aaedcf5a749d56bfe2fa80d04348fe00
02d5ea1cec65004ee3522855837aac88b2f4cbab11eb5b0d84dbf2159ba83eef6b726d0a02a0f16b
07cf26dd522bc34f5c1a5f871bcca6c3ab76c05a7b1392a627afb4ed0501f4791d13dda9c7c5eeaf
8d6d054d43c88d4645b684430e8e2b88e661bef9ce118c466c7e17d03a4078ef723ea91b40221095
f8a1b4edfdc199ae34fdcaee1c79e8e475f9544aea53958582a3351535265ae346dd5705a273e619
76c084c1f221712fd779a473c3cb8666708f94d4d67501d73b39b1e0a525b53c4f93dceadde58798
1acde2e8a0208d1b5eb397880e279074ef318185d7169a04c215fda5d940743dadf8d

Original Q
Q = 8d28148505135d7d6a1ab5075553e3516a3e0eea9158926eded37eca42a5e321

Generated Q
gen Q = 961daaebfc7fac7ffe3cca30e6bcd699c0f72ffdc1bf152b3bd5e7b721ed33fb

Original counter:
228

Generated counter:
235

**Comment by smuellerDD** [comment]
After looking and testing some more, I found that the generated values constantly change. For OpenSSL, this is a sign that hash(domain_seed) does not produce a prime number. In this case, OpenSSL simply generates a new random number for the domain seed and retries the PQ generation operation with the new seed.

Note, OpenSSL uses a M-R primality test with 64 rounds and in addition an attempt to divide the value to be tested with the first 2048 primes (2, 3, 5, ...).

**Comment by celic** [comment]
ACVP also uses a M-R primality test with a high number of rounds... in this case (L = 2048) 56 rounds. It is extremely unlikely to find a value that passes M-R for so many rounds then fails after a few more rounds. Can you tell if it is the M-R that fails for OpenSSL or if it is the further check you mentioned? I'll poke around with the value the server provided on our end. But this is very curious...

**Comment by celic** [comment]
Also `hash(domain_seed)` is not necessarily prime I don't think. The specification (FIPS 186-4) states that...
```
U = Hash (domain_parameter_seed) mod 2^(N–1).
q = 2^(N–1) + U + 1 – ( U mod 2).
Then test whether or not q is prime as specified in Appendix C.3...
```
From what you are saying it seems like the server provided `q` is not a prime (and hence OpenSSL is falling back to it's normal routines to ensure it finds a prime `q`). Is this correct?

**Comment by atvassilev** [comment]
@smuellerDD , since we are discussing primality testing for OpenSSL, I think you may find interesting this recent paper on the subject: https://eprint.iacr.org/2018/749.pdf.

**Comment by celic** [comment]
While we don't have a deterministic primality test implemented... as far as my testing can see, the `q` the server provided is in fact prime and should verify from the given parameters.

**Comment by celic** [comment]
Ah I see the problem. The `domain_seed` the server provides is 248 bits when all these values should be 256 bits. This is likely because the first 8 bits are all 0s. So the hex for `domain_seed` should be `"00D1F806D6651B7718A2B38F6C5FEE17273475789AF5AACA242AE1F5A5C2DFC9"`. This could affect the client trying to hash the seed due to the missing bits. The server should be stricter at enforcing this length however. @smuellerDD please let me know if this works if you can test it quickly.

**Comment by smuellerDD** [comment]
That is the trick. After adding a right-pad to the data, it passes.

Thanks a lot!

**Comment by celic** [comment]
This is still a server error. It has been fixed internally for v0.5 but will be left alone for v0.4.
